### ISSUE TITLE
snscrape not working giving ScraperException error

### ISSUE THREAD

**Issue by tasnimaziz** [JustAnotherArchivist/snscrape#247]
ScraperException: 4 requests to https://api.twitter.com/2/search/adaptive.json?include_profile_interstitial_type=1&include_blocking=1&include_blocked_by=1&include_followed_by=1&include_want_retweets=1&include_mute_edge=1&include_can_dm=1&include_can_media_tag=1&skip_status=1&cards_platform=Web-12&include_cards=1&include_ext_alt_text=true&include_quote_count=true&include_reply_count=1&tweet_mode=extended&include_entities=true&include_user_entities=true&include_ext_media_color=true&include_ext_media_availability=true&send_error_codes=true&simple_quoted_tweets=true&q=from%3AAITCofficial&tweet_search_mode=live&count=100&query_source=spelling_expansion_revert_click&pc=1&spelling_corrections=1&ext=mediaStats%2ChighlightedLabel failed, giving up.

**Comment by JustAnotherArchivist** [comment]
Unable to reproduce, `snscrape twitter-user AITCofficial` works fine here. Please provide more details about the error and the environment.

**Comment by tasnimaziz** [comment]
thanks...now its working

**Comment by eyaler** [comment]
got this now for the first time
```
Error retrieving https://api.twitter.com/2/search/adaptive.json?include_profile_interstitial_type=1&include_blocking=1&include_blocked_by=1&include_followed_by=1&include_want_retweets=1&include_mute_edge=1&include_can_dm=1&include_can_media_tag=1&skip_status=1&cards_platform=Web-12&include_cards=1&include_ext_alt_text=true&include_quote_count=true&include_reply_count=1&tweet_mode=extended&include_entities=true&include_user_entities=true&include_ext_media_color=true&include_ext_media_availability=true&send_error_codes=true&simple_quoted_tweets=true&q=lang%3Ahe&tweet_search_mode=live&count=100&query_source=spelling_expansion_revert_click&cursor=scroll%3AthGAVUV0VFVBaGgL3lh63usicWgIC5pd6bj7snEnEV46vzCBWAiXoYB0RFRkFVTFQ1ARW6oQEVAAA%3D&pc=1&spelling_corrections=1&ext=mediaStats%2ChighlightedLabel: non-200 status code
4 requests to https://api.twitter.com/2/search/adaptive.json?include_profile_interstitial_type=1&include_blocking=1&include_blocked_by=1&include_followed_by=1&include_want_retweets=1&include_mute_edge=1&include_can_dm=1&include_can_media_tag=1&skip_status=1&cards_platform=Web-12&include_cards=1&include_ext_alt_text=true&include_quote_count=true&include_reply_count=1&tweet_mode=extended&include_entities=true&include_user_entities=true&include_ext_media_color=true&include_ext_media_availability=true&send_error_codes=true&simple_quoted_tweets=true&q=lang%3Ahe&tweet_search_mode=live&count=100&query_source=spelling_expansion_revert_click&cursor=scroll%3AthGAVUV0VFVBaGgL3lh63usicWgIC5pd6bj7snEnEV46vzCBWAiXoYB0RFRkFVTFQ1ARW6oQEVAAA%3D&pc=1&spelling_corrections=1&ext=mediaStats%2ChighlightedLabel failed, giving up.
Traceback (most recent call last):
  File "t.py", line 47, in <module>
    for tweet in sntwitter.TwitterSearchScraper('lang:'+args.lang+maxid_arg).get_items():
  File "/home/ubuntu/anaconda3/envs/py38/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 521, in get_items
    for obj in self._iter_api_data('https://api.twitter.com/2/search/adaptive.json', params, paginationParams):
  File "/home/ubuntu/anaconda3/envs/py38/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 232, in _iter_api_data
    obj = self._get_api_data(endpoint, reqParams)
  File "/home/ubuntu/anaconda3/envs/py38/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 202, in _get_api_data
    r = self._get(endpoint, params = params, headers = self._apiHeaders, responseOkCallback = self._check_api_response)
  File "/home/ubuntu/anaconda3/envs/py38/lib/python3.8/site-packages/snscrape/base.py", line 199, in _get
    return self._request('GET', *args, **kwargs)
  File "/home/ubuntu/anaconda3/envs/py38/lib/python3.8/site-packages/snscrape/base.py", line 195, in _request
    raise ScraperException(msg)
snscrape.base.ScraperException: 4 requests to https://api.twitter.com/2/search/adaptive.json?include_profile_interstitial_type=1&include_blocking=1&include_blocked_by=1&include_followed_by=1&include_want_retweets=1&include_mute_edge=1&include_can_dm=1&include_can_media_tag=1&skip_status=1&cards_platform=Web-12&include_cards=1&include_ext_alt_text=true&include_quote_count=true&include_reply_count=1&tweet_mode=extended&include_entities=true&include_user_entities=true&include_ext_media_color=true&include_ext_media_availability=true&send_error_codes=true&simple_quoted_tweets=true&q=lang%3Ahe&tweet_search_mode=live&count=100&query_source=spelling_expansion_revert_click&cursor=scroll%3AthGAVUV0VFVBaGgL3lh63usicWgIC5pd6bj7snEnEV46vzCBWAiXoYB0RFRkFVTFQ1ARW6oQEVAAA%3D&pc=1&spelling_corrections=1&ext=mediaStats%2ChighlightedLabel failed, giving up.
```

**Comment by JustAnotherArchivist** [comment]
@eyaler Hmm, interesting, also unable to reproduce. `snscrape --verbose twitter-search --cursor 'scroll:thGAVUV0VFVBaGgL3lh63usicWgIC5pd6bj7snEnEV46vzCBWAiXoYB0RFRkFVTFQ1ARW6oQEVAAA=' 'lang:he'` works fine for me on f9a3fafb. Do you have a dump file?

**Comment by eyaler** [comment]
@JustAnotherArchivist thanks for your quick support. my report was before @f9a3faf, but just got this again with that patch as well (also, could you kindly explain - was the cursor ignored before?). anyway i am not sure how to call sntwitter.TwitterSearchScraper with vv and --dump-locals

**Comment by eyaler** [comment]
third time: i ran `snscrape --verbose --dump-locals twitter-search 'lang:he' &> tweets.log`
log ends with:
```
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/envs/py38/bin/snscrape", line 8, in <module>
    sys.exit(main())
  File "/home/ubuntu/anaconda3/envs/py38/lib/python3.8/site-packages/snscrape/_cli.py", line 280, in main
    for i, item in enumerate(scraper.get_items(), start = 1):
  File "/home/ubuntu/anaconda3/envs/py38/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 521, in get_items
    for obj in self._iter_api_data('https://api.twitter.com/2/search/adaptive.json', params, paginationParams, cursor = self._cursor):
  File "/home/ubuntu/anaconda3/envs/py38/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 232, in _iter_api_data
    obj = self._get_api_data(endpoint, reqParams)
  File "/home/ubuntu/anaconda3/envs/py38/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 202, in _get_api_data
    r = self._get(endpoint, params = params, headers = self._apiHeaders, responseOkCallback = self._check_api_response)
  File "/home/ubuntu/anaconda3/envs/py38/lib/python3.8/site-packages/snscrape/base.py", line 199, in _get
    return self._request('GET', *args, **kwargs)
  File "/home/ubuntu/anaconda3/envs/py38/lib/python3.8/site-packages/snscrape/base.py", line 195, in _request
    raise ScraperException(msg)
snscrape.base.ScraperException: 4 requests to https://api.twitter.com/2/search/adaptive.json?include_profile_interstitial_type=1&include_blocking=1&include_blocked_by=1&include_followed_by=1&include_want_retweets=1&include_mute_edge=1&include_can_dm=1&include_can_media_tag=1&skip_status=1&cards_platform=Web-12&include_cards=1&include_ext_alt_text=true&include_quote_count=true&include_reply_count=1&tweet_mode=extended&include_entities=true&include_user_entities=true&include_ext_media_color=true&include_ext_media_availability=true&send_error_codes=true&simple_quoted_tweets=true&q=lang%3Ahe&tweet_search_mode=live&count=100&query_source=spelling_expansion_revert_click&cursor=scroll%3AthGAVUV0VFVBaAgLLVgZmIsicWjsC-0aadvLwnEnEVh9zvCxWAiXoYB0RFRkFVTFQ1ARWS0gEVAAA%3D&pc=1&spelling_corrections=1&ext=mediaStats%2ChighlightedLabel failed, giving up.
```
i am not sure if a dump file was created and i did not find one.

**Comment by JustAnotherArchivist** [comment]
Yeah, `--verbose`, `--dump-locals`, and the automatic dumping of local variables on crashes are part of the CLI implementation , so they're not available when using the scrapers directly. `--verbose` can be done though as snscrape simply uses the standard `logging` module; it simply calls `logging.getLogger().setLevel(logging.INFO)` (or `DEBUG` if specified more than once) in the CLI.

There should be a `FATAL`-level log message `Dumped stack and locals to $FILENAME` just before the traceback. The file should be called `snscrape_locals_*` and be located in your `TMPDIR`.

**Comment by eyaler** [comment]
did not find the message but found the files:
[snsscrape.zip](https://github.com/JustAnotherArchivist/snscrape/files/6930664/snsscrape.zip)

**Comment by JustAnotherArchivist** [comment]
Thanks, interesting. How quickly after starting the scrape does this happen? If you start another scrape immediately after one crashes, does that work? What sort of connection are you scraping from (residential, cloud server, etc.)?

**Comment by eyaler** [comment]
interesting how? it happens after some time... let me know what can help and i can put in some measurements. i usually waited at least a few minutes after failing and it was ok. i am on an EC2. but this is a recent issue started only a few days ago and now happening every time

**Comment by JustAnotherArchivist** [comment]
Interesting because it's an HTTP 403 response that causes the crash. I haven't seen that mid-scrape before. Also interesting that you're able to scrape from EC2 at all. I was under the impression that AWS IPs were banned entirely, cf. #79.

You could try this patch:

```patch
diff --git a/snscrape/modules/twitter.py b/snscrape/modules/twitter.py
index 07b5b36..9d9461e 100644
--- a/snscrape/modules/twitter.py
+++ b/snscrape/modules/twitter.py
@@ -187,10 +187,10 @@ class TwitterAPIScraper(snscrape.base.Scraper):
                del self._apiHeaders['x-guest-token']
 
        def _check_api_response(self, r):
-               if r.status_code == 429:
+               if r.status_code in (429, 403):
                        self._unset_guest_token()
                        self._ensure_guest_token()
-                       return False, 'rate-limited'
+                       return False, 'rate-limited' if r.status_code == 429 else 'forbidden'
                if r.headers.get('content-type', '').replace(' ', '') != 'application/json;charset=utf-8':
                        return False, 'content type is not JSON'
                if r.status_code != 200:
```

If that fixes it, you should see an `INFO` log message containing `forbidden` and then a resumption of the scraping after refetching the search page.

**Comment by gauravbinjola312** [comment]
Error retrieving https://api.twitter.com/2/search/adaptive.json?include_profile_interstitial_type=1&include_blocking=1&include_blocked_by=1&include_followed_by=1&include_want_retweets=1&include_mute_edge=1&include_can_dm=1&include_can_media_tag=1&skip_status=1&cards_platform=Web-12&include_cards=1&include_ext_alt_text=true&include_quote_count=true&include_reply_count=1&tweet_mode=extended&include_entities=true&include_user_entities=true&include_ext_media_color=true&include_ext_media_availability=true&send_error_codes=true&simple_quoted_tweets=true&q=%28from%3AIyappanMLA%29+until%3A2022-11-01+since%3A2022-10-01&tweet_search_mode=live&count=100&query_source=spelling_expansion_revert_click&pc=1&spelling_corrections=1&ext=mediaStats%2ChighlightedLabel: non-200 status code
4 requests to https://api.twitter.com/2/search/adaptive.json?include_profile_interstitial_type=1&include_blocking=1&include_blocked_by=1&include_followed_by=1&include_want_retweets=1&include_mute_edge=1&include_can_dm=1&include_can_media_tag=1&skip_status=1&cards_platform=Web-12&include_cards=1&include_ext_alt_text=true&include_quote_count=true&include_reply_count=1&tweet_mode=extended&include_entities=true&include_user_entities=true&include_ext_media_color=true&include_ext_media_availability=true&send_error_codes=true&simple_quoted_tweets=true&q=%28from%3AIyappanMLA%29+until%3A2022-11-01+since%3A2022-10-01&tweet_search_mode=live&count=100&query_source=spelling_expansion_revert_click&pc=1&spelling_corrections=1&ext=mediaStats%2ChighlightedLabel failed, giving up.
Traceback (most recent call last):
  File "/home/ubuntu/scrap_social/Gaurav_dev/twitter/new_twt_monthly.py", line 77, in <module>
    for tweet in sntwitter.TwitterSearchScraper(query).get_items():
  File "/home/ubuntu/scrap_social/scrap-env/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 680, in get_items
    for obj in self._iter_api_data('https://api.twitter.com/2/search/adaptive.json', params, paginationParams, cursor = self._cursor):
  File "/home/ubuntu/scrap_social/scrap-env/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 369, in _iter_api_data
    obj = self._get_api_data(endpoint, reqParams)
  File "/home/ubuntu/scrap_social/scrap-env/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 339, in _get_api_data
    r = self._get(endpoint, params = params, headers = self._apiHeaders, responseOkCallback = self._check_api_response)
  File "/home/ubuntu/scrap_social/scrap-env/lib/python3.8/site-packages/snscrape/base.py", line 216, in _get
    return self._request('GET', *args, **kwargs)
  File "/home/ubuntu/scrap_social/scrap-env/lib/python3.8/site-packages/snscrape/base.py", line 212, in _request
    raise ScraperException(msg)
snscrape.base.ScraperException: 4 requests to https://api.twitter.com/2/search/adaptive.json?include_profile_interstitial_type=1&include_blocking=1&include_blocked_by=1&include_followed_by=1&include_want_retweets=1&include_mute_edge=1&include_can_dm=1&include_can_media_tag=1&skip_status=1&cards_platform=Web-12&include_cards=1&include_ext_alt_text=true&include_quote_count=true&include_reply_count=1&tweet_mode=extended&include_entities=true&include_user_entities=true&include_ext_media_color=true&include_ext_media_availability=true&send_error_codes=true&simple_quoted_tweets=true&q=%28from%3AIyappanMLA%29+until%3A2022-11-01+since%3A2022-10-01&tweet_search_mode=live&count=100&query_source=spelling_expansion_revert_click&pc=1&spelling_corrections=1&ext=mediaStats%2ChighlightedLabel failed, giving up.
![Screenshot (9)](https://user-images.githubusercontent.com/88122610/201029740-db0ee866-c608-4945-ae13-f802b471188c.png)

**Comment by gauravbinjola312** [comment]
So we can't scrape data for these accounts

**Comment by TheTechRobo** [comment]
> You can work around it by adding `(filter:safe OR -filter:safe)` to the search query.

**Comment by aakashb95** [comment]
TLDR: try for 400 Bad Request try adding `-filter:unsafe` to your query

We were using this code to get user data. And it worked fine even for users that would return non-200 status code.
But for a specific user 'EhsanullahIrf10', inspite of catching ScraperException, python code would close.

```python
  tweets_df = pd.DataFrame(columns=['datetime', 'tweet_id', 'text', "tweet_location",
'username',"like_count", "reply_count","retweet_count", "last_scraped_date",
"sentiment","status", "linked_keyword",
"sentiment_score","tweet_uid"])
  try:
        # Using TwitterSearchScraper to scrape data and append tweets to list
        for i,tweet in enumerate(sntwitter.TwitterSearchScraper(
            f"(from:{username}) since:{time_frame}").get_items()
            ):
            '''
            if i>maxTweets:
            break
            '''
            tweet_date = tweet.date.strftime('%Y-%m-%d %H:%M:%S')
            tweet_uid = f"{tweet.id}{keyword}"
            tweets_list.append([tweet_date, tweet.id, tweet.content,tweet.coordinates,
                                tweet.user.username,tweet.likeCount,
                                tweet.replyCount,tweet.retweetCount,
                                current_time, "NA","I",keyword,0.0,tweet_uid])

        tweets_df = pd.DataFrame(tweets_list, columns=['datetime', 'tweet_id', 'text', 
        "tweet_location",'username',"like_count","reply_count",
        "retweet_count","last_scraped_date","sentiment","status","linked_keyword","sentiment_score","tweet_uid"]
        )
  except ScraperException:
        return tweets_df
```

So we tried using CLI:
```
snscrape --jsonl --progress --max-results 4 --since 2022-07-07 twitter-search "(from:EhsanullahIrf10) since:2022-07-07" > ehsan.json
```

Twitter advanced search also said Incorrect query. 
![image](https://user-images.githubusercontent.com/48802744/211157574-a421ae75-e86d-4d9c-ac23-9370fe4352cc.png)

Funnily enough, this is inconsistent and works for other users.
We checked the /tmp/ file generated from CLI and got ` r.status_code = 400`

We tried your solution
> More specifically, that's an HTTP 400 response. I encountered that for the first time a couple days ago. It's a bug on Twitter's side and happens reproducibly for certain accounts (mine was `askDrIra`). You can work around it by adding `(filter:safe OR -filter:safe)` to the search query.

-filter: safe resulted in unwanted results for us

an example:
```
snscrape --jsonl --progress --max-results 4 --since 2022-07-07 twitter-search "(from:EhsanullahIrf10) since:2022-07-07 -filter:safe" > ehsan-safe.json
```

Gave us 4 objects where usernames were not EhsanullahIrf10. (one sample had "username": "oge323")

We checked their twitter and they don't seem to have any tweets since 2022-07-07 so ideally we should have an empty json. 

Unsure of what to do, we simply tried 
`-filter:unsafe` :smile: 

AND it seems to work for this account. No idea how or why.

```
snscrape --jsonl --progress --max-results 4 --since 2022-07-07 twitter-search "(from:EhsanullahIrf10) since:2022-07-07 -filter:unsafe" > ehsan-unsafe.json
```

results in 0 records and twitter advanced search also says no results
![image](https://user-images.githubusercontent.com/48802744/211157533-4d128224-9ef1-4633-81bc-996dd181810d.png)

**Comment by JustAnotherArchivist** [comment]
@aakashb95 `-filter:unsafe` might lead to missing tweets. That's why my suggestion is `(filter:safe OR -filter:safe)`. Those aren't two options; add the entire thing to the search query. It acts as a no-op filter and should in theory not affect the results at all.

    snscrape twitter-search 'from:EhsanullahIrf10 (filter:safe OR -filter:safe)'

I can reproduce the weird results with just `-filter:safe` though. No idea what that's about. Twitter being Twitter, I guess. It only happens in combination with `since:2022-07-07`; removing that returns the expected results.

**Comment by annanasi-mon** [comment]
Hi, my code seemed to work previously very well and now when I run it I get the following error message:

`Error retrieving https://api.twitter.com/2/search/adaptive.json?include_profile_interstitial_type=1&include_blocking=1&include_blocked_by=1&include_followed_by=1&include_want_retweets=1&include_mute_edge=1&include_can_dm=1&include_can_media_tag=1&skip_status=1&cards_platform=Web-12&include_cards=1&include_ext_alt_text=true&include_quote_count=true&include_reply_count=1&tweet_mode=extended&include_entities=true&include_user_entities=true&include_ext_media_color=true&include_ext_media_availability=true&send_error_codes=true&simple_quoted_tweets=true&q=%22%23bitcoin+since%3A2021-01-01+until%3A2021-12-31+lang%3Aen%22&tweet_search_mode=live&count=100&query_source=spelling_expansion_revert_click&pc=1&spelling_corrections=1&ext=mediaStats%2ChighlightedLabel: non-200 status code
4 requests to https://api.twitter.com/2/search/adaptive.json?include_profile_interstitial_type=1&include_blocking=1&include_blocked_by=1&include_followed_by=1&include_want_retweets=1&include_mute_edge=1&include_can_dm=1&include_can_media_tag=1&skip_status=1&cards_platform=Web-12&include_cards=1&include_ext_alt_text=true&include_quote_count=true&include_reply_count=1&tweet_mode=extended&include_entities=true&include_user_entities=true&include_ext_media_color=true&include_ext_media_availability=true&send_error_codes=true&simple_quoted_tweets=true&q=%22%23bitcoin+since%3A2021-01-01+until%3A2021-12-31+lang%3Aen%22&tweet_search_mode=live&count=100&query_source=spelling_expansion_revert_click&pc=1&spelling_corrections=1&ext=mediaStats%2ChighlightedLabel failed, giving up.
Traceback (most recent call last):
  File "/snap/pycharm-community/312/plugins/python-ce/helpers/pydev/pydevconsole.py", line 364, in runcode
    coro = func()
  File "<input>", line 49, in <module>
  File "/home/anna/PycharmProjects/Master_thesis/venv/lib/python3.10/site-packages/pandas/core/frame.py", line 736, in __init__
    data = list(data)
  File "/home/anna/PycharmProjects/Master_thesis/venv/lib/python3.10/site-packages/snscrape/modules/twitter.py", line 680, in get_items
    for obj in self._iter_api_data('https://api.twitter.com/2/search/adaptive.json', params, paginationParams, cursor = self._cursor):
  File "/home/anna/PycharmProjects/Master_thesis/venv/lib/python3.10/site-packages/snscrape/modules/twitter.py", line 369, in _iter_api_data
    obj = self._get_api_data(endpoint, reqParams)
  File "/home/anna/PycharmProjects/Master_thesis/venv/lib/python3.10/site-packages/snscrape/modules/twitter.py", line 339, in _get_api_data
    r = self._get(endpoint, params = params, headers = self._apiHeaders, responseOkCallback = self._check_api_response)
  File "/home/anna/PycharmProjects/Master_thesis/venv/lib/python3.10/site-packages/snscrape/base.py", line 216, in _get
    return self._request('GET', *args, **kwargs)
  File "/home/anna/PycharmProjects/Master_thesis/venv/lib/python3.10/site-packages/snscrape/base.py", line 212, in _request
    raise ScraperException(msg)
snscrape.base.ScraperException: 4 requests to https://api.twitter.com/2/search/adaptive.json?include_profile_interstitial_type=1&include_blocking=1&include_blocked_by=1&include_followed_by=1&include_want_retweets=1&include_mute_edge=1&include_can_dm=1&include_can_media_tag=1&skip_status=1&cards_platform=Web-12&include_cards=1&include_ext_alt_text=true&include_quote_count=true&include_reply_count=1&tweet_mode=extended&include_entities=true&include_user_entities=true&include_ext_media_color=true&include_ext_media_availability=true&send_error_codes=true&simple_quoted_tweets=true&q=%22%23bitcoin+since%3A2021-01-01+until%3A2021-12-31+lang%3Aen%22&tweet_search_mode=live&count=100&query_source=spelling_expansion_revert_click&pc=1&spelling_corrections=1&ext=mediaStats%2ChighlightedLabel failed, giving up.` 





And here is the code:

`data_sentiment = pd.DataFrame(itertools.islice(sntwitter.TwitterSearchScraper(
    '"#bitcoin since:2021-01-01 until:2021-12-31 lang:en"').get_items(), 100)) 
`

**Comment by Yaman-yagiz** [comment]
I have same problem, also i tried everything but again i got same result. I have exactly same error output...

**Comment by JustAnotherArchivist** [comment]
This is an old issue and entirely unrelated to the current problem, which is tracked in #634 (pinned and near the top of the issue list!).
